c10/core/impl/TorchDispatchModeTLS.cpp

```diff
@@ -8,22 +7,26 @@ namespace impl {
 
 thread_local TorchDispatchModeTLS torchDispatchModeState;
 
-void TorchDispatchModeTLS::push_onto_stack(std::shared_ptr<SafePyObject> mode) {
+void TorchDispatchModeTLS::push_onto_stack(
+    std::shared_ptr<c10::ModePyObjTrampoline> mode) {
   if (torchDispatchModeState.stack_.size() == 0) {
     c10::impl::tls_set_dispatch_key_included(DispatchKey::Python, true);
     c10::impl::tls_set_dispatch_key_included(
         DispatchKey::PythonTLSSnapshot, true);
   }
+  mode->mode_state_push_trampoline();
```

Comment:
This and L29 are exactly equivalent of what Voz did in the original PR. This just occurs in the C++ so that all push/pop calls are noted (this lets us properly handle, say, reentrant dispatch)